DSharpPlus/Net/RestClient.cs

```diff
@@ -14,27 +15,35 @@
 
 namespace DSharpPlus.Net
 {
+    /// <summary>
+    /// Creates an instance of a Http client implementation.
+    /// </summary>
+    /// <param name="messageHandler">Message handler to use for the new Http client instance.</param>
+    /// <returns>Constructed Http client implementation.</returns>
+    public delegate HttpClient HttpClientFactoryDelegate(HttpMessageHandler messageHandler);
+
     /// <summary>
     /// Represents a client used to make REST requests.
     /// </summary>
     internal sealed class RestClient
     {
         private static UTF8Encoding UTF8 { get; } = new UTF8Encoding(false);
         private static Regex RouteArgumentRegex { get; } = new Regex(@":([a-z_]+)");
+        private static TimeSpan RateLimitResetBuffer { get; } = TimeSpan.FromMilliseconds(10);
 
         private BaseDiscordClient Discord { get; }
         private HttpClient HttpClient { get; }
-        private HashSet<RateLimitBucket> Buckets { get; }
-        private SemaphoreSlim RequestSemaphore { get; }
+        private ConcurrentDictionary<string, RateLimitBucket> Buckets { get; }
```

Comment:
You're saying this has the potential to throw, and throwing can cause deadlocks. If you're going with this approach, I would request that you make sure nothing around this throws.

Reply:
I meant that the current implementation using the HashSet can throw when multiple threads make requests and the set is modified while the other thread still searches for its Bucket using the ConcurrentDictionary avoids this.

Reply:
A'ight, that's fair.